src/libRumorSpreading/RumorHolder.cpp

```diff
@@ -0,0 +1,400 @@
+    RumorHolder::receivedMessage(const Message& message, int fromPeer)
+    {
+        std::lock_guard<std::mutex> guard(m_mutex); // critical section
+
+        bool isNewPeer = m_peersInCurrentRound.insert(fromPeer).second;
+        increaseStatValue(StatisticKey::NumMessagesReceived, 1);
+
+        // If this is the first time 'fromPeer' sent a PUSH/EMPTY_PUSH message in this round
+        // then respond with a PULL message for each rumor
+        std::vector<Message> pullMessages;
+        if (isNewPeer
+            && (message.type() == Message::Type::PUSH
+                || message.type() == Message::Type::EMPTY_PUSH))
+        {
+            for (auto& kv : m_rumors)
+            {
+                RumorStateMachine& stateMach = kv.second;
+                if (stateMach.age() > 0 and !stateMach.isOld())
+                {
+                    pullMessages.emplace_back(Message(
+                        Message::Type::PULL, kv.first, kv.second.age()));
+                }
+            }
+
+            // No PULL messages to sent i.e. no rumors received yet,
+            // then send EMPTY_PULL Message. so the receiving peer will
+            if (pullMessages.empty())
+            {
+                pullMessages.emplace_back(
+                    Message(Message::Type::EMPTY_PULL, -1, 0));
+                increaseStatValue(StatisticKey::NumEmptyPullMessages, 1);
+            }
+            else
+            {
+                increaseStatValue(StatisticKey::NumPullMessages,
+                                  pullMessages.size());
+                m_nonPriorPeers.insert(fromPeer);
+            }
+        }
+
+        // An empty response from a peer that was sent a PULL
+        const int receivedRumorId = message.rumorId();
+        const int theirRound = message.age();
+        if (receivedRumorId >= 0)
+        {
+            if (m_rumors.count(receivedRumorId) > 0)
+            {
+                m_rumors.at(receivedRumorId)
+                    .rumorReceived(fromPeer, message.age());
+            }
+            else
+            {
+                m_rumors.insert(std::make_pair(
+                    receivedRumorId,
+                    RumorStateMachine(&m_networkConfig, fromPeer, theirRound)));
+            }
+        }
+
+        return std::make_pair(fromPeer, pullMessages);
+    }
+
+    std::pair<std::vector<int>, std::vector<Message>>
+    RumorHolder::advanceRound()
+    {
+        std::lock_guard<std::mutex> guard(m_mutex); // critical section
+
+        if (m_peers.size() == 0)
+        {
+            m_nonPriorPeers.clear();
+            m_peersInCurrentRound.clear();
+            return std::make_pair(std::vector<int>{-1}, std::vector<Message>());
+        }
+
+        increaseStatValue(StatisticKey::Rounds, 1);
+
+        std::vector<int> toMembers;
+        int toMember;
+        int retryCount = 0;
+        int neighborC = 0;
+
+        if (m_nonPriorPeers.size() >= m_peers.size() - m_maxNeighborsPerRounds)
```

Comment:
I think `m_nonPriorPeers` actually means the list of peers that have already been used before? If that's the case, shouldn't it be `m_priorPeers`? Regardless, I suggest a more descriptive name, like `m_peersUsedInPrevRounds`. Also, please add some comments where this is used in the code below.

Reply:
Oops I just realized Prior means priority, not prior.  In that case, definitely change the name please.